crates/llvm-context/src/polkavm/context/mod.rs

```diff
@@ -471,15 +485,133 @@ where
+        &self,
+        name: &str,
+        line_no: u32,
+    ) -> anyhow::Result<inkwell::debug_info::DISubprogram<'ctx>> {
+        let dinfo = self.debug_info().expect("expected debug-info builder");
+        let di_scope = dinfo.top_scope().expect("expected a debug-info scope");
+        let di_builder = dinfo.builder();
+        let di_file = dinfo.compilation_unit().get_file();
+        let di_flags = inkwell::debug_info::DIFlagsConstants::PUBLIC;
+        let ret_type = dinfo.create_word_type(Some(di_flags))?.as_type();
+        let subroutine_type =
+            di_builder.create_subroutine_type(di_file, Some(ret_type), &[], di_flags);
+        let linkage = self
+            .debug_info()
+            .expect("expected debug-info builders")
+            .namespace_as_identifier(Some(name));
+        let func_scope = di_builder.create_function(
+            di_scope,
+            name,
+            Some(linkage.as_str()),
+            di_file,
+            line_no,
+            subroutine_type,
+            false,
+            true,
+            1,
+            di_flags,
+            false,
+        );
+        Ok(func_scope)
+    }
+
+    /// Adds a debug-info scope to the current function.
+    pub fn set_current_function_debug_info(
+        &mut self,
+        name: &str,
+        line: usize,
+    ) -> anyhow::Result<inkwell::debug_info::DISubprogram<'ctx>> {
+        let line_num: u32 = std::cmp::min(line, u32::MAX as usize) as u32;
+        let func_value = self
+            .current_function()
+            .borrow()
+            .declaration()
+            .function_value();
+        let func_scope = match func_value.get_subprogram() {
+            Some(scp) => scp,
+            None => self.build_function_debug_info(name, line_num)?,
+        };
+        func_value.set_subprogram(func_scope);
+        self.set_debug_location(line, 0, Some(func_scope.as_debug_info_scope()))?;
+        Ok(func_scope)
+    }
+
+    pub fn set_debug_location(
+        &self,
+        line: usize,
+        column: usize,
+        scope: Option<DIScope<'ctx>>,
+    ) -> anyhow::Result<()> {
+        if let Some(dinfo) = self.debug_info() {
+            let line_num: u32 = std::cmp::min(line, u32::MAX as usize) as u32;
+            let column_num: u32 = std::cmp::min(column, u32::MAX as usize) as u32;
+            let di_scope = match scope {
+                Some(scp) => scp,
+                None => dinfo.top_scope().expect("expected a debug-info scope"),
+            };
+            let di_loc = dinfo.builder().create_debug_location(
+                self.llvm(),
+                line_num,
+                column_num,
+                di_scope,
+                None,
+            );
+            self.builder().set_current_debug_location(di_loc);
+        }
         Ok(())
     }
 
+    /// Pushes a debug-info scope to the stack.
+    pub fn push_debug_scope(&self, scope: DIScope<'ctx>) {
+        if self.debug_info().is_some() {
```

Comment:
Should prefer `if let` instead of `is_some()` followed by unwraps when we operate on optional values only if they are `Some` (noticed this in other places too).